### PR TITLE
NEP17 example

### PR DESCRIPTION
**Pull request by chenzhitong**

### CODE REVIEW

File: csharp/NEP17/NEP17.Helpers.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+using Neo;
+using Neo.SmartContract.Framework;
+using Neo.SmartContract.Framework.Services.Neo;
+
+namespace Template.NEP17.CSharp
+{
+    public partial class NEP17 : SmartContract
+    {
+        private static bool ValidateAddress(UInt160 address) => address.IsValid && !address.IsZero;
+        private static bool IsDeployed(UInt160 address) => ManagementContract.GetContract(address) != null;
```
Comment (erikzhang) [inline csharp/NEP17/NEP17.Helpers.cs:10]: Did you use the latest code? I think it is the old code which has the `ManagementContract`.

File: csharp/NEP17/NEP17.csproj [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
```
Comment (erikzhang) [inline csharp/NEP17/NEP17.csproj:4]: `net5.0`?

**Comment by chenzhitong** [comment]
I replaced `Neo.SmartContract.Framework 3.0.0-preview4` with the latest dll, because preview4 does not include `ContractManagement` and compile failure.

**Review by vncoelho — COMMENTED** [review]
Maybe we need to delete all old templates in this repository and create a master2x branch, it will be confusing if we merge just this one that will be up-to-date.

How about these other two examples here https://github.com/neo-project/neo-devpack-dotnet/tree/master/templates?

**Comment by erikzhang** [comment]
> I replaced Neo.SmartContract.Framework 3.0.0-preview4 with the latest dll, because preview4 does not include ContractManagement and compile failure.

Noooo. Please revert it.

**Comment by chenzhitong** [comment]
> Maybe we need to delete all old templates in this repository and create a master2x branch, it will be confusing if we merge just this one that will be up-to-date.

@Tommo-L  agrees.

**Comment by erikzhang** [comment]
> Maybe we need to delete all old templates in this repository

The old examples should be updated.

> How about these other two examples here https://github.com/neo-project/neo-devpack-dotnet/tree/master/templates?

They are not examples. They are vs templates.

**Comment by Tommo-L** [comment]
It's better to update these old examples after preview5.

>  create a master2x branch, it will be confusing if we merge just this one that will be up-to-date.

Indeed, we need to create a master-2.x branch before merge this pr.

**Comment by superboyiii** [comment]
We need create a master-2.x branch and we should update these old examples in preview5.

**Comment by erikzhang** [comment]
No master-2.x branch is needed. We are focusing on 3.0.

**Comment by vncoelho** [comment]
Thus we need to delete all projects in this repository, @erikzhang.
Otherwise it will be confusing.

I prefer to keep these examples on DevPack for now.

**Comment by superboyiii** [comment]
Let's move on.

**Comment by superboyiii** [comment]
@erikzhang Can we using git workflow on devpack to update smartcontract.framwork to MyGet? Otherwise, this exmaple will not compiled successfully unless using devpack project reference.

**Comment by erikzhang** [comment]
Maybe we can update this repo for each preview.

**Comment by vncoelho** [comment]
Sounds good,@erikzhang, if we make a release for each preview with some couple of examples.

**Comment by erikzhang** [comment]
Please revert the last commit.

File: csharp/NEP17/NEP17.Methods.cs [hunk 3]
```diff
@@ -31,7 +32,7 @@ public static bool Transfer(UInt160 from, UInt160 to, BigInteger amount, object
             OnTransfer(from, to, amount);
 
             // Validate payable
-            if (IsDeployed(to)) Contract.Call(to, "onPayment", new object[] { from, amount, data });
+            if (IsDeployed(to)) Contract.Call(to, "onPayment", CallFlags.ReadOnly, new object[] { from, amount, data });
```
Comment (shargon) [inline csharp/NEP17/NEP17.Methods.cs]:
```suggestion
            if (IsDeployed(to)) Contract.Call(to, "onNEP17Payment", CallFlags.ReadOnly, new object[] { from, amount, data });
```

**Comment by erikzhang** [comment]
Wait for neo-project/neo-devpack-dotnet#461

**Comment by chenzhitong** [comment]
Reviewe?

File: csharp/NEP17/NEP17.Owner.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+namespace Neo.SmartContract.Examples
+{
+    partial class NEP17Demo
+    {
+        public static void _deploy(object data, bool update)
+        {
+            if (update) return;
+            if (TotalSupplyStorage.Get() > 0) throw new Exception("Contract has been deployed.");
+
+            TotalSupplyStorage.Increase(InitialSupply);
```
Comment (shargon) [inline csharp/NEP17/NEP17.Owner.cs:14]: Use mint?

File: csharp/NEP17/NEP17.Methods.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            if (from == to) return true;
+
+            AssetStorage.Reduce(from, amount);
+            AssetStorage.Increase(to, amount);
+
+            OnTransfer(from, to, amount);
+
+            // Validate payable
+            if (ContractManagement.GetContract(to) != null)
+                Contract.Call(to, "onNEP17Payment", CallFlags.ReadOnly, new object[] { from, amount, data });
```
Comment (shargon) [inline csharp/NEP17/NEP17.Methods.cs:28]: `ReadOnly`?

File: csharp/NEP17/NEP17.Owner.cs (hunk 4 above)
Comment (chenzhitong) [reply csharp/NEP17/NEP17.Owner.cs:14]: What does it mean?

File: csharp/NEP17/NEP17.Methods.cs (hunk 5 above)
Comment (chenzhitong) [reply csharp/NEP17/NEP17.Methods.cs:28]: What do you think is the best permission to give?

File: csharp/NEP17/NEP17.Owner.cs (hunk 4 above)
Comment (erikzhang) [reply csharp/NEP17/NEP17.Owner.cs:14]: @shargon `Mint` is for crowd sale I think.

File: csharp/NEP17/NEP17.Methods.cs (hunk 5 above)
Comment (erikzhang) [reply csharp/NEP17/NEP17.Methods.cs:28]: I think it should be `CallFlags.All`.

**Comment by superboyiii** [comment]
@erikzhang @chenzhitong @shargon Let's complete this before RC1.

File: csharp/NEP17/NEP17.cs [hunk 6]
```diff
@@ -10,6 +10,8 @@ namespace Neo.SmartContract.Examples
     [ManifestExtra("Description", "This is a NEP17 example")]
     [SupportedStandards("NEP-17")]
     [ContractPermission("*", "onNEP17Payment")]
+    [ContractPermission("0xfffdc93764dbaddd97c48f252a53ea4643faa3fd", "update")]
```
Comment (shargon) [inline csharp/NEP17/NEP17.cs:13]: `*`?

Comment (erikzhang) [reply csharp/NEP17/NEP17.cs:13]: For static calls, the compiler should know which permissions to add.

Comment (shargon) [reply csharp/NEP17/NEP17.cs:13]: I will change the compiler

Comment (chenzhitong) [reply csharp/NEP17/NEP17.cs:13]: OK, I revert this commit

**Comment by superboyiii** [comment]
Test **PASS**
Works well on mine.
@shargon @erikzhang  Need your review.

**Review by erikzhang — APPROVED** [review]

**Comment by shargon** [comment]
Give me one hour

**Review by shargon — DISMISSED** [review]

**Review by erikzhang — APPROVED** [review]

**Review by shargon — APPROVED** [review]

**Review by vncoelho — APPROVED** [review]

**Pull request merged**